torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,485 @@
+                if rank == self.rank:
+                    # Send the state to the reference replica
+                    dist.broadcast_object_list([local_cpu_state], src=self.global_rank, group=self.group)
+                else:
+                    global_rank = _get_global_rank(self.group, rank)
+
+                    # Discard this tensor/rank, broadcast necessary for syncing and because NCCL does not support gather
+                    dist.broadcast_object_list([0], src=global_rank, group=self.group)
+
+    def partition_parameters(self) -> List[List[Dict]]:
+        """Partitions parameters across distributed data parallel ranks.
+
+        Returns: a list of ``param_groups`` (which is a list of dict) where each
+        element of the list contains the param_groups for a rank. Element 0
+        corresponds to rank 0, etc. We need all the ranks for the broadcast
+        inside ``step()``.
+        """
+        if len(self._partition_parameters) == 0:
+            self._partition_parameters = [list() for _ in range(self.world_size)]
+            sizes = [0] * self.world_size
+            for param_group in self.param_groups:
+                param_lists: List[List] = [list() for _ in range(self.world_size)]
+                for param in param_group["params"]:
+                    # Add this param to rank with smallest size.
+                    rank = sizes.index(min(sizes))
+                    param_lists[rank].append(param)
+                    sizes[rank] += param.numel()
+
+                for rank, params in enumerate(param_lists):
+                    param_group_rank = copy.copy(param_group)
+                    param_group_rank["params"] = params
+                    self._partition_parameters[rank].append(param_group_rank)
+
+        return self._partition_parameters
+
+    @property
+    def per_device_params(self) -> Dict[torch.device, List[List[Parameter]]]:
+        """Sorted list of all the params, first per device then per rank.
+
+        Within a list params are sorted per number of elements to allow for an easy bucketing.
+        """
+        if len(self._per_device_params) == 0:
+            # Go through all params, log them per device
+            # The ordering is important here, needs to be the same on all ranks
+            # So that ulterior broadcast calls are matching
+            for param_group in self.param_groups:
+                for param in param_group["params"]:
+                    device = param.device
+                    if self._per_device_params.get(device) is None:
+                        self._per_device_params[device] = [[] for _ in range(self.world_size)]
+                    self._per_device_params[device][self.param_to_rank[param]] += [param]
+
+            # Sort param_lists by size
+            for k in self._per_device_params.keys():
+                for r in self._per_device_params[k]:
+                    r.sort(key=lambda x: x.numel())
```

Comment:
if there are equal sizes, is it guaranteed that all ranks will still see the same order?

Reply:
ohh, good point ! I'm guessing that it follows the insertion order in that case, or something similar, at least if the python interpreter is the same everywhere then the same sub-rule should apply ?

Reply:
I would assume so, but if it is not, it will be very hard to debug. Because very likely, it won't crash (as sizes are the same), but communication can yield wrong result I guess?

Reply:
I just checked, the [doc](https://docs.python.org/3/library/stdtypes.html#list.sort) mentions that

> The sort() method is guaranteed to be stable. A sort is stable if it guarantees not to change the relative order of elements that compare equal

in that case it should keep initial insertion order which will be the same for all ranks (well, depending on pytorch but if that changed then everything would be broken anyway I think)